src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -32,48 +32,53 @@ protected override void EndProcessing()
         {
             string filePath = null;
             string tempPath = Path.GetTempPath();
-            int attempts = 0;
-            bool temporaryFilePathFound = false;
-            while(attempts++ < 10 && !temporaryFilePathFound)
-            {
-                string fileName = Path.GetRandomFileName();
-                fileName = Path.ChangeExtension(fileName, _extension);
-                filePath = Path.Combine(tempPath, fileName);
-                if(!File.Exists(filePath))
-                {
-                    temporaryFilePathFound = true;
-                }
-            }
-            if(!temporaryFilePathFound)
-            {
-                WriteError(
-                    new ErrorRecord(
-                            new IOException("Could not find an available temporary file name in {tempPath}."),
-                            "NewTemporaryFileResourceUnavailable",
-                            ErrorCategory.ResourceUnavailable,
-                            tempPath));
-                    return;
-            }
 
-            if (ShouldProcess(filePath))
+            if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    using (new FileStream(filePath, FileMode.CreateNew)) { }
-                    FileInfo file = new FileInfo(filePath);
-                    WriteVerbose("Created temporary file {file.FullName}.");
-                    WriteObject(file);
+                    try
+                    {
+                        string fileName = Path.GetRandomFileName();
+                        fileName = Path.ChangeExtension(fileName, _extension);
+                        filePath = Path.Combine(tempPath, fileName);
+                        using (new FileStream(filePath, FileMode.CreateNew)) { }
+                        creationOfFileSuccessful = true;
+                        WriteVerbose("Created temporary file {filePath}.");
+                    }
+                    catch (IOException) // file already exists -> retry
+                    {
+                        attempts++;
+                    }
+                    catch (Exception exception) // fatal error, which could be e.g. insufficient permissions, etc.
+                    {
+                        WriteError(
+                        new ErrorRecord(
+                                exception,
+                                "NewTemporaryFileWriteError",
+                                ErrorCategory.WriteError,
+                                tempPath));
+                        return;
+                    }
                 }
-                catch (Exception e)
+
+                if (!creationOfFileSuccessful)
                 {
                     WriteError(
```

Comment:
Understood, @PaulHigin. 

@bergmeister: Just to clarify:  by default, the script would continue to run even if a terminating error were reported with `ThrowTerminatingError()`, because only the _statement_ is terminated; the difference between reporting a non-terminating vs. a terminating error only matters with respect to `-ErrorAction` use and `try` / `catch` use.

Since the conceptual fog around error handling lifted for me personally only recently, perhaps you'll find this useful too: [Our Error Handling, Ourselves - time to fully understand and properly document PowerShell's error handling](https://github.com/PowerShell/PowerShell-Docs/issues/1583)